plugins/obs-ffmpeg/texture-amf.cpp

```diff
@@ -0,0 +1,1129 @@
+				      encoder_packet *packet)
+{
+	if (!data)
+		return;
+
+	enc->packet_data = AMFBufferPtr(data);
+	data->GetProperty(L"PTS", &packet->pts);
+
+	bool hevc = enc->codec == amf_codec_type::HEVC;
+	const wchar_t *get_output_type =
+		hevc ? AMF_VIDEO_ENCODER_HEVC_OUTPUT_DATA_TYPE
+		     : AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE;
+
+	uint64_t type;
+	data->GetProperty(get_output_type, &type);
+
+	switch (type) {
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR:
+		packet->priority = OBS_NAL_PRIORITY_HIGHEST;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_I:
+		packet->priority = OBS_NAL_PRIORITY_HIGH;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_P:
+		packet->priority = OBS_NAL_PRIORITY_LOW;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_B:
+		packet->priority = OBS_NAL_PRIORITY_DISPOSABLE;
+		break;
+	}
+
+	packet->data = (uint8_t *)enc->packet_data->GetNative();
+	packet->size = enc->packet_data->GetSize();
+	packet->type = OBS_ENCODER_VIDEO;
+	packet->dts = convert_to_obs_ts(enc, data->GetPts());
+	packet->keyframe = type == AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR;
+}
+
+static bool amf_encode_tex(void *data, uint32_t handle, int64_t pts,
+			   uint64_t lock_key, uint64_t *next_key,
+			   encoder_packet *packet, bool *received_packet)
+try {
+	amf_data *enc = (amf_data *)data;
+	ID3D11DeviceContext *context = enc->context;
+	ComPtr<ID3D11Texture2D> output_tex;
+	ComPtr<ID3D11Texture2D> input_tex;
+	ComPtr<IDXGIKeyedMutex> km;
+	AMFSurfacePtr amf_surf;
+	AMFDataPtr amf_out;
+	AMF_RESULT res;
+
+	if (handle == GS_INVALID_HANDLE) {
+		*next_key = lock_key;
+		throw "Encode failed: bad texture handle";
+	}
+
+	get_tex_from_handle(enc, handle, &km, &input_tex);
+	if (!enc->texture)
+		create_texture(enc, input_tex);
+	output_tex = enc->texture;
+
+	/* ------------------------------------ */
+	/* copy to output tex                   */
+
+	km->AcquireSync(lock_key, INFINITE);
+	context->CopyResource((ID3D11Resource *)output_tex.Get(),
+			      (ID3D11Resource *)input_tex.Get());
+	context->Flush();
+	km->ReleaseSync(*next_key);
+
+	/* ------------------------------------ */
+	/* map output tex to amf surface        */
+
+	/* XXX: does this really need to be recreated every frame? */
+
+	res = enc->amf_context->CreateSurfaceFromDX11Native(output_tex,
+							    &amf_surf, nullptr);
+	if (res != AMF_OK)
+		throw amf_error("CreateSurfaceFromDX11Native failed", res);
+
+	int64_t last_ts = convert_to_amf_ts(enc, pts - 1);
+	int64_t cur_ts = convert_to_amf_ts(enc, pts);
+
+	amf_surf->SetPts(cur_ts);
+	amf_surf->SetProperty(L"PTS", pts);
+	amf_surf->SetDuration(cur_ts - last_ts);
+
+	/* ------------------------------------ */
+	/* do actual encode                     */
+
+	res = enc->amf_encoder->SubmitInput(amf_surf);
+	if (res != AMF_OK)
+		throw amf_error("SubmitInput failed", res);
+
+	for (;;) {
+		res = enc->amf_encoder->QueryOutput(&amf_out);
+		if (res == AMF_OK)
+			break;
+
+		switch (res) {
+		case AMF_NEED_MORE_INPUT:
```

Comment:
QueryOutput cannot return AMF_NEED_MORE_INPUT.

Reply:
This is how it was originally written in #4538 -- I admit I did not catch that.